juju/client/connection.py

```diff
@@ -22,49 +23,73 @@
-    'ModelGeneration': {'versions': [4]},
-    'ModelManager': {'versions': [9]},
-    'ModelUpgrader': {'versions': [1]},
-    'Payloads': {'versions': [1]},
-    'Pinger': {'versions': [1]},
-    'Resources': {'versions': [3]},
-    'SSHClient': {'versions': [4]},
-    'SecretBackends': {'versions': [1]},
-    'Secrets': {'versions': [1, 2]},
-    'Spaces': {'versions': [6]},
-    'Storage': {'versions': [6]},
-    'Subnets': {'versions': [5]},
-    'UserManager': {'versions': [3]},
+    'Action': [7],
+    'Admin': [3],
+    'AllModelWatcher': [4],
+    'AllWatcher': [3],
+    'Annotations': [2],
+    'Application': [17, 19],
+    'ApplicationOffers': [4],
+    'Backups': [3],
+    'Block': [2],
+    'Bundle': [6],
+    'Charms': [6],
+    'Client': [6, 7],
+    'Cloud': [7],
+    'Controller': [11],
+    'CredentialManager': [1],
+    'FirewallRules': [1],
+    'HighAvailability': [2],
+    'ImageMetadataManager': [1],
+    'KeyManager': [1],
+    'MachineManager': [10],
+    'MetricsDebug': [2],
+    'ModelConfig': [3],
+    'ModelGeneration': [4],
+    'ModelManager': [9],
+    'ModelUpgrader': [1],
+    'Payloads': [1],
+    'Pinger': [1],
+    'Resources': [3],
+    'SSHClient': [4],
+    'SecretBackends': [1],
+    'Secrets': [1, 2],
+    'Spaces': [6],
+    'Storage': [6],
+    'Subnets': [5],
+    'UserManager': [3],
 }
 
+# We don't generate code for these, as we can never use them.
+# The controller happily lists them though, don't warn about these.
+known_unsupported_facades = frozenset(("""
+ActionPruner Agent AgentLifeFlag AgentTools ApplicationScaler CAASAdmission CAASAgent
+CAASApplication CAASApplicationProvisioner CAASFirewaller CAASFirewallerSidecar
+CAASModelConfigManager CAASModelOperator CAASOperator CAASOperatorProvisioner
+CAASOperatorUpgrader CAASUnitProvisioner CharmDownloader CharmRevisionUpdater Cleaner
+CredentialValidator CrossModelRelations CrossModelSecrets Deployer DiskManager
+EntityWatcher EnvironUpgrader ExternalControllerUpdater FanConfigurer
+FilesystemAttachmentsWatcher Firewaller HostKeyReporter ImageMetadata InstanceMutater
+InstancePoller KeyUpdater LeadershipService LifeFlag LogForwarding Logger MachineActions
+MachineUndertaker Machiner MeterStatus MetricsAdder MetricsManager MigrationFlag
+MigrationMaster MigrationMinion MigrationStatusWatcher NotifyWatcher OfferStatusWatcher
+PayloadsHookContext Provisioner ProxyUpdater Reboot RelationStatusWatcher
+RelationUnitsWatcher RemoteRelationWatcher RemoteRelations ResourcesHookContext
+RetryStrategy SecretBackendsManager SecretBackendsRotateWatcher SecretsDrain
+SecretsManager SecretsRevisionWatcher SecretsTriggerWatcher Singular StatusHistory
+StorageProvisioner StringsWatcher Undertaker UnitAssigner Uniter UpgradeSeries
+UpgradeSteps Upgrader UserSecretsDrain UserSecretsManager VolumeAttachmentPlansWatcher
+VolumeAttachmentsWatcher
+""" + """
+CrossController MigrationTarget ModelSummaryWatcher NotifyWatcher
+""").strip().split())
```

Comment:
@james-garner-canonical adding these here to suppress these warnings that we now have in main:

```

2024-10-17T04:25:45.1538548Z WARNING  juju.client.connection:connection.py:844 unexpected facade CAASOperatorProvisioner received from the controller
2024-10-17T04:25:45.1539858Z WARNING  juju.client.connection:connection.py:844 unexpected facade CAASOperatorUpgrader received from the controller
2024-10-17T04:25:45.1541273Z WARNING  juju.client.connection:connection.py:844 unexpected facade CAASUnitProvisioner received from the controller
2024-10-17T04:25:45.1542700Z WARNING  juju.client.connection:connection.py:844 unexpected facade CharmDownloader received from the controller
2024-10-17T04:25:45.1544309Z WARNING  juju.client.connection:connection.py:844 unexpected facade CharmRevisionUpdater received from the controller
2024-10-17T04:25:45.1546190Z WARNING  juju.client.connection:connection.py:844 unexpected facade Cleaner received from the controller
2024-10-17T04:25:45.1547670Z WARNING  juju.client.connection:connection.py:844 unexpected facade CredentialValidator received from the controller
2024-10-17T04:25:45.1549404Z WARNING  juju.client.connection:connection.py:844 unexpected facade CrossModelRelations received from the controller
2024-10-17T04:25:45.1550709Z WARNING  juju.client.connection:connection.py:844 unexpected facade CrossModelSecrets received from the controller
... etc.
```

Source: https://productionresultssa6.blob.core.windows.net/actions-results/2e4c880c-4537-4404-8298-85eb503413dc/workflow-job-run-59585bb8-35b7-5c33-d6e9-30fd13ff141d/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-10-17T04%3A54%3A15Z&sig=FIcaec1iZpJeYonhmnXoqasej%2FVJC76pSfRQe4b4EC4%3D&ske=2024-10-17T16%3A28%3A47Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-10-17T04%3A28%3A47Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-08-04&sp=r&spr=https&sr=b&st=2024-10-17T04%3A44%3A10Z&sv=2024-08-04

Reply:
I didn't realise that Juju sends us facades too, but I see how they're used in the facade version negotiation logic. Makes sense for us to explicitly filter out the known non-client ones. Maybe we can look into if there's a way to only ask for the client ones, or file an issue with Juju requesting such a feature in future.

That link doesn't seem to work for me btw.